rust/connlib/tunnel/src/tests/composite_strategy.rs

```diff
@@ -0,0 +1,104 @@
+use proptest::strategy::{BoxedStrategy, Strategy, Union};
+use std::fmt;
+
+#[derive(Debug)]
+pub struct CompositeStrategy<T>
+where
+    T: fmt::Debug,
+{
+    options: Vec<(u32, BoxedStrategy<T>)>,
+}
+
+impl<T> Default for CompositeStrategy<T>
+where
+    T: fmt::Debug,
+{
+    fn default() -> Self {
+        Self {
+            options: Vec::default(),
+        }
+    }
+}
+
+impl<T> CompositeStrategy<T>
+where
+    T: fmt::Debug,
+{
+    /// Adds a strategy to this [`CompositeStrategy`].
+    pub fn with(mut self, prob: u32, strategy: impl Strategy<Value = T> + 'static) -> Self {
+        self.options.push((prob, strategy.boxed()));
+
+        self
+    }
+
+    /// Adds a strategy based on some input element if the element is not empty.
+    pub fn with_if_not_empty<S, E>(
+        self,
+        prob: u32,
+        element: E,
+        make_strategy: impl Fn(E) -> S,
+    ) -> Self
+    where
+        S: Strategy<Value = T> + 'static,
+        E: IsEmpty,
+    {
+        if element.is_empty() {
+            return self;
+        }
+
+        self.with(prob, make_strategy(element))
+    }
+}
+
+pub trait IsEmpty {
+    fn is_empty(&self) -> bool;
+}
```

Comment:
this might be a silly idea, but what if instead of `IsEmpty` we require `Default` and we consider `x` to be empty iff `x == Default::defatult()`?

Reply:
Default isn't implemented on tuples so I wouldn't be able to pass a tuple of elements in then which is required for the DNS query transition.

Reply:
I thought it was implemented for all `T` that implement default? https://doc.rust-lang.org/std/primitive.tuple.html#impl-Default-for-(T,)

Reply:
Ah lol, I checked the page for `Default` and it didn't show up there. Maybe I missed it!

I do think it is a bit backwards though but feel free to change it if you want :)